components/javascript-wrapper.js

```diff
@@ -0,0 +1,157 @@
+var rpf = require('./rpf');
+
+/**
+ * This module provides a Javascript wrapper for updaters.  It is responsible for setting up the
+ * code to ensure updater callback, and then invoking the RDF pipeline framework to complete building
+ * the component and setting up the vni and state metadata.
+ *
+ * @param nodeDef updater node definition. This can be a simple updater function, or an object that contains 
+ *        updater metadata (e.g., input/output port definitions, description, icons), and an updater.
+ *
+ * @return a promise to create the noflo rdf component
+ */
+module.exports = function(nodeDef) { 
+
+    // Save the caller's node instance 
+    var self = this;
+    
+    // Ensure we have at least one input port and and updater, using a default stub updater if 
+    // there is none specified
+    var updaterArgs = null;
+    if ( _.isUndefined(nodeDef) || 
+       ( ! _.isFunction(nodeDef) && ( _.isEmpty( nodeDef ) || ( _.isEmpty( nodeDef.inPorts ))))) {
+       // Got nothing to work with - no input ports and no updater, so create a default updater and input port 
+       nodeDef = { 
+           inPorts: {
+               input: {
+                   datatype: 'object',
+                   description: "default input port" 
+               }
+           },
+           updater: defaultUpdater
+       };
+
+    }  else if ( _.isUndefined( nodeDef.updater ) && !_.isFunction(nodeDef)) { 
+       // Got some node metadata, but no node updater - use the metadata with default updater
+       nodeDef = _.extend( nodeDef, 
+                           { updater: defaultUpdater } );
+
+    } else if ( _.isFunction( nodeDef ) ) { 
+        // Got an updater but no metadata - use updater args as input port list
+        // This assumes that all updater arguments are input
+        updaterArgs = introspect( nodeDef );
+        nodeDef = { inPorts: updaterArgs,
+                    updater: nodeDef };
+    } 
+
+    // Have an updater now, no matter what, be it default or not - figure out what the updater arguments are
+    var updaterArgs = (_.isNull( updaterArgs )) ? introspect( nodeDef.updater ) : updaterArgs; 
+    var inPorts = _.isArray(nodeDef.inPorts) ? _.object(nodeDef.inPorts, []) : nodeDef.inPorts;
+
+    return promiseComponent( 
+                         _.defaults({
+                             inPorts: _.mapObject(inPorts, function(port, name) {
+                                 return _.extend( port, 
+                                                  { ondata: ondata.bind( self, name, updaterArgs, nodeDef )});
+                             })
+                           }, nodeDef)
+                     );
+
+}; // module.exports
+
+// Define a default updater stub function to be used if the caller 
+// did not pass one in.  Simply returns whatever the original input was so it
+// can be sent through the output port to the next component
+var defaultUpdater = function() { 
+    return { out: arguments };
```

Comment:
It looks like line 51 does #1 if no inPorts were specified, but l think it would be good to do it even if some inPorts were specified.  In other words, for any updater parameter (other than 'output') that is not already listed in inPorts, add it to inPorts.

For #2, yes I now see that line 35 does that already.

For the return, the question is how we want users to be able to write their updaters.  I'm thinking that it would be simpler for users (since a node now can only have one output state) if the updater could just write:
<pre>
return newData;
</pre>
instead of
<pre>
return {output: newData}
</pre>
but we can discuss this.  The downside of returning newData directly is that we still need some way for the updater to indicate that the state did not change, such as:
<pre>
return rpf.NO_CHANGE;
</pre>
That doesn't look very pretty, but it is explicit.  See slide 6, which I just updated:
http://dbooth.org/2013/pipeline/RDF-Pipeline-Implementation.pdf